### DISCUSSION TITLE
Creating multiple tables with read_json based on data

### DISCUSSION THREAD

**Discussion by v1gnesh** [duckdb/duckdb#9195] [Q&A]
Hello!

Say I create a single table with this:
```sql
create table bigtable as select * from read_json('out.ndjson.zstd', format='newline_delimited', auto_detect='true', compression='zstd');
```
and that there's a struct named `header` in the JSON source (and therefore in `bigtable` after the import).
The `header` struct has a field (called `typ`) to indicate the type of the particular JSON object.

I would like to do something like:

```sql
create table {typ-table} as select * from read_json('out.ndjson.zstd', format='newline_delimited', auto_detect='true', compression='zstd') where header.typ = {each-unique-typ};
```

Is something like this possible today, if so, how do I do this?
It's fine if it involves multiple statements, as long as I don't have to manually get the unique `typ` list and create them as tables.

In short, it would be good to have partitioning based on some JSON/struct key, to save into different tables.
The reverse of hive partitioning (where partitioned data is read into a single table).

**Comment by l1t1** [comment]
see recursive example in https://github.com/duckdb/duckdb/issues/9169

**Reply by v1gnesh** [reply]
I'm not sure I follow. There's nothing to unnest (recursively) in my case...?

**Comment by l1t1** [comment]
please post your sample data file.

**Reply by v1gnesh** [reply]
```json
{"header": {"a": 1, "b": 2, "typ": 3}, "body": Typ3}
{"header": {"a": 1, "b": 2, "typ": 23}, "body": Typ23}
{"header": {"a": 1, "b": 2, "typ": 5}, "body": Typ5}
```

**Comment by Mause** [comment]
There's nothing stopping you from doing:

```sql
create table main_table as select * from read_json('out.ndjson.zstd', format='newline_delimited', auto_detect='true', compression='zstd') where header.typ = {each-unique-typ};
create table sub_table as select * from main_table where header.typ = {each-unique-typ};
```

?

**Reply by Mause** [reply]
You just can't presently automatically create a new table for each sub type (without resorting to generating SQL)

**Reply by v1gnesh** [reply]
Yeah, it gets "tiring" to do it manually for hundreds of `typ`s.
Thinking if I should do something like 
- print unique list of `header.typ` into a list in python
- iterate over list to create `create table` commands

But just wondering if there's a way to filter in-line as I'm doing `read_json`.
Filtering is possible, but only to select specific columns, not value-based... right?

**Comment by l1t1** [comment]
make a sql script by 
```
create table t as select distinct  header.typ  x from main_table;
select 'create table sub_table_' || x ||' as select * from main_table where header.typ ='''  || x || ''''from t;
```

**Reply by v1gnesh** [reply]
Thank you, I see the commands are generated:

```sql
create table st_125 as select * from main_table where header.typ='125'
create table st_154 as select * from main_table where header.typ='154'
```

I'm not very familiar with SQL at all. Is there a way to directly pipe these as commands?
Would just putting the commands you've given in a SQL script and executing execute the above commands that it produces?

**Comment by l1t1** [comment]
use example, ref http://duckdb.org/docs/archive/0.9.0/api/cli
```
~ $ vi a.sql
~ $ cat a.sql
.mode c
.headers off
select 'select '||(42+i)||' a;' from range(3) t(i)
~ $ duckdb < a.sql | duckdb
┌───────┐
│   a   │
│ int32 │
├───────┤
│    42 │
└───────┘
┌───────┐
│   a   │
│ int32 │
├───────┤
│    43 │
└───────┘
┌───────┐
│   a   │
│ int32 │
├───────┤
│    44 │
└───────┘

```
a more parctical example, copy tpch test data to parquet files
```
.mode c
.headers off
load tpch;
select 'load tpch;';
CALL dbgen(sf=0.01);
select 'CALL dbgen(sf=0.01);';
select 'copy '||table_name||' to ''/tmp/'||table_name||'.parquet'';'from information_schema.tables;
```

**Reply by v1gnesh** [reply]
TBH, I don't understand this at all...

**Comment by l1t1** [comment]
let first duckdb run the a.sql script to generate SQL statements, then pipe them to the second duckdb to run.
